### ISSUE TITLE
Monitor/Cron checkin attachment upload reports 404

### ISSUE THREAD

**Issue by Radiergummi** [getsentry/sentry#56247]
### Environment

SaaS (https://sentry.io/)

### Steps to Reproduce

We attempt to upload backup logs alongside failed checkins. According to the logs, this should work like so:

```bash
sentry_upload_logs() {
    local slug="${SENTRY_MONITOR_SLUG}-${1}"
    local log_file="$(mktemp)"
    fetch_logs > "${log_file}"

    curl --request "POST" "https://sentry.io/api/0/organizations/${SENTRY_ORG}/monitors/${slug}/checkins/latest/attachment/" \
        --header "Authorization: DSN https://${SENTRY_PUBLIC_KEY}@${SENTRY_INGEST_HOST}/${SENTRY_PROJECT_ID}" \
        --form "file=${log_file}"

    rm -f "${log_file}"
}

sentry_upload_logs "full"
```

**Edit:** adding more detail.

We have a backup job running monthly (full), weekly (diff), and daily (incremental); for all three, we need monitors, aptly named `database-backup-{full,diff,incr}`. To make reusing the setup easy, I used the Upsert feature to create/update the monitor with the first `in_progress` checkin in the backup script after starting. Then, if it encounters any errors, the script will send its own journald logs to  its monitor attachment endpoint on Sentry, and only after that send the final checkin with the `error` status (everything as recommended in the docs).  
I can provide the full script as necessary, but it doesn't really contain anything different from the documentation.

### Expected Result

A checkin with an attachment visible in Sentry; no errors.

### Actual Result

The server responds with a 404: `{"detail":"The requested resource does not exist"}`.

I also tried to use a locally generated checkin-ID instead of `latest`, but that yields the same result. And, I tried replacing "monitors" with "crons" in the URL (the docs use both, which is a bit confusing), but that just yielded a 404 without a body, so I assume that route just doesn't exist in your backend.

### Product Area

Crons

### Link

https://sentry.io/api/0/organizations/matchory/monitors/pipeline-database-backup-full/checkins/latest/attachment/

### DSN

https://{PKEY}@o444794.ingest.sentry.io/4504241492721664

### Version

-

**Comment by Radiergummi** [comment]
Er... this got a bit stranger now :) Instead of a *missing* resource, the endpoint now reports the following: `{"detail":"Check-in already has an attachment"}`. 

I've tried this both with `latest` and a randomly generated UUID. What is going on here?

**Comment by Radiergummi** [comment]
Okay, I think I've figured it out. There appears to be some race conditions on your end. What works for me is the following order:

1. Upsert the monitor, set status to `in_progress`
2. Create another checkin right after, setting the status to `in_progress`
3. Perform the work
4. On failure, sleep for at least two seconds
5. Upload the attachment
6. Do the final checkin, setting the status to `error`

Here's the script for posterity:
<details>

<pre lang="bash">

sentry_checkin() {
    local slug="${SENTRY_MONITOR_SLUG}-${1}"
    local check_in_id="${3:-$(uuidgen)}"

    if [ -n "${3}" ]; then
        curl "https://${SENTRY_INGEST_HOST}/api/${SENTRY_PROJECT_ID}/cron/${slug}/${SENTRY_PUBLIC_KEY}/?status=${2}" \
            --silent
    else
        curl --request POST \
            "https://${SENTRY_INGEST_HOST}/api/${SENTRY_PROJECT_ID}/cron/${slug}/${SENTRY_PUBLIC_KEY}/" \
            --header 'Content-Type: application/json' \
            --data-raw "${2}" \
            --silent
    fi

    echo "${check_in_id}"
}

sentry_upload_logs() {
    local slug="${SENTRY_MONITOR_SLUG}-${1}"
    local check_in_id="${2}"
    local temp_directory
    temp_directory="$(mktemp -d)"
    local log_file="${temp_directory}/${slug}-${START}.log"

    fetch_logs > "${log_file}"

    curl --request "POST" \
        "https://sentry.io/api/0/organizations/${SENTRY_ORG}/monitors/${slug}/checkins/latest/attachment/" \
        --header "Authorization: DSN https://${SENTRY_PUBLIC_KEY}@${SENTRY_INGEST_HOST}/${SENTRY_PROJECT_ID}" \
        --form "file=@${log_file}" \
        --no-progress-meter

    rm -rf "${temp_directory}"
}

check_in_id=$(sentry_checkin "${type}" "{
  \"monitor_config\": {
  \"schedule\": {
    \"type\": \"interval\",
    \"value\": 1,
    \"unit\": \"${unit}\"
  }
    },
    \"status\": \"in_progress\"
}")

check_in_id=$(sentry_checkin "${type}" "in_progress" "${check_in_id}")

if backup "${type}"; then
    sentry_checkin "${type}" "ok" "${check_in_id}" > /dev/null
    exit 0
else
    sleep 2
    sentry_upload_logs "${type}" "${check_in_id}"
    sentry_checkin "${type}" "error" "${check_in_id}" > /dev/null
    exit 1
fi

</pre>

</details>

----

Both the duplicate checkins and sleep appear to be necessary if the job fails very fast. If I had to guess I'd say the `latest` checkin reference isn't updated fast enough (or async).  I tried using my client-generated checkin ID, but that would always result in the resource being not found.

**Comment by rjo100** [comment]
Hi! Thanks for the report.

Right now `latest` isn't intended to be used for the `attachments` endpoint as its primary usage is for signaling a finished status for an `in-progress` check-in. When we're looking up the latest check-in, we're excluding all but ones that are `in-progress`. That would explain the behavior where it's only working when you create a second check-in.

Since you already are passing the `check_in_id` to the `upload_logs` function, maybe try `/checkins/${check_in_id}/attachments/` instead of `/checkins/latest/attachment/`. That should directly reference the check-in and successfully upload the attachment.

In the meantime, we'll update the docs to more clearly reflect that `latest` only means in-progress and also look into making the behavior line up more with expectations.

Let us know if you have any more questions.

**Comment by Radiergummi** [comment]
Hi, thank you for the response!

> That would explain the behavior where it's only working when you create a second check-in.

Are you sure? Because the second check-in is also created as `in_progress`, so by those rules it probably shouldn't pass for `latest` either...

> Since you already are passing the `check_in_id` to the `upload_logs` function, maybe try `/checkins/${check_in_id}/attachments/` instead of `/checkins/latest/attachment/`. That should directly reference the check-in and successfully upload the attachment.

That was what I originally tried, but then the API would simply respond with a 404:
```json
{"detail":"The requested resource does not exist"}
```

The docs are a bit confusing, because depending on the scenario, you'll need both the cron ingestion endpoint (authenticated implicitly via the URL), and the Sentry API (authenticated via DSN in a header), which both work a bit differently.  
Actually, I just realised the entire attachments section seems to be... gone? I swear it was there when I set up our backup :(

**Comment by rjo100** [comment]
Sorry if I wasn't clear, `latest` will only access the most recent `in_progress` check-in. For attachments, you need to reference a completed check-in id.

For the `check-in id`, the `id` in the URL should match exactly for both API calls. If you don't generate an id to send with the check-in, we return one from the API endpoint when creating the check-in that you should also use for the attachment endpoint.

Sorry about the docs, we're still in beta and are working on moving some things around on our backend to make things simpler long term.

And as for the UI, if there's no attachments we just shipped a small improvement to hide the column. Apologies for the confusion!

**Comment by gaprl** [comment]
Hey @Radiergummi, thanks for reaching out and inquiring about attachments. 👋 

I wanted to provide additional context on Crons' support for attachments. In our [latest update](https://docs.sentry.io/product/crons/legacy-endpoint-migration/), we've decided to remove support for uploading attachments directly through our API. This is primarely due to these endpoints not being designed to handle ingestion-level traffic, and as we get closer to GA, we needed to focus our efforts elsewhere.

We understand this is not ideal. However, we do provide a different solution that could work well for your scenario. We've recently added support for tracing errors during a check-in, and since you mentioned attachments are helpful when errors occur, I recommend taking a look at the [attachments documentation](https://docs.sentry.io/platforms/python/enriching-events/attachments/) for issues.

Here's how this can work:

1. Use one of our [supported SDKs](https://docs.sentry.io/product/crons/getting-started/) to manage your check-ins.
2. Activate error monitoring for your application.
3. Upon detecting an issue during your runtime, you can create a new Sentry issue with an attachment.
4. The issue will be linked to the check-in and will be shown in your monitor details page. Here's an example:

![CleanShot 2023-09-19 at 12 27 14@2x](https://github.com/getsentry/sentry/assets/1308394/661299dd-1e5e-4b25-8ebb-7853b8030f15)

I understand this might not work for your case since your example was posted in bash. In any case, we haven't completely ruled out supporting attachments again in the future, and will definetely take this feedback into consideration when planning our roadmap.

Thanks for your understanding, and do let us know if the suggestion above could work for you.

**Comment by Radiergummi** [comment]
To be honest, attachments *did* feel a little out of place. I'm probably not going to rewrite the backup script in a proper language -- it's just a thin wrapper around pgBackRest, and 95% of it are Sentry incantations anyway -- but I'm pretty sure this is going to be helpful to others in a more, say, app-y scenario.

However, what I think would be an actually better solution would be key-value metadata on the check-ins. I'm not particularly attached (heh) to adding the entire service log as an attachment, if I can simply add the error reported by the command I'm running (in this case, pgBackRest).

Thank you for all the detailed explanations on this issue. Really appreciate it!

**Comment by gaprl** [comment]
Thank you @Radiergummi, adding key-value metadata support for check-ins is definetely higher up on our list. The method I described above can work in the same way by using [context](https://docs.sentry.io/platforms/python/enriching-events/context/) instead of attachments.

I'll close this issue for now, if you have any other problems feel free to reopen or submit a new issue. Thanks!
